baseapp/baseapp.go

```diff
@@ -540,12 +535,27 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 		ctx = ctx.WithMultiStore(msCache)
 	}
 
-	result = handler(ctx, msg)
+	for _, msg := range msgs {
+		// Match route.
+		msgType := msg.Type()
+		handler := app.router.Route(msgType)
+		if handler == nil {
+			return sdk.ErrUnknownRequest("Unrecognized Msg type: " + msgType).Result()
+		}
 
-	// Set gas utilized
-	result.GasUsed = ctx.GasMeter().GasConsumed()
+		result = handler(ctx, msg)
```

Comment:
We need to combine the results in some reasonably intelligent manner:

- Return all codes / data / logs / info / tags
- Add up gas used, return total used at the end

Reply:
Good point.

Seems like we can just append data and logs together, possibly with some seperator in between individual msg data/logs.

Tags and ValidatorUpdates should also be appended but dropping redundant elements.

Where is GasWanted set? Do we have to change its behavior with multiple msgs?

The code should only be nonzero if the msg fails, right?

Maybe we should do this:

Since we abort running the tx on the first message that fails, the result on a failed transaction should just be the result of the first failed message.

In the case where the tx successfully executes, we accumulate data/logs/info/tags in the way described above and return the result with code 0.

I think this makes it much easier to parse a failed result of a transaction since the user only cares about which message caused the failure.

Thoughts?

Reply:
> Seems like we can just append data and logs together, possibly with some seperator in between individual msg data/logs.

I think the invariant *always* needs to hold that running a multi-Msg transaction has the same effect on the state as running the Msgs individually in separate transactions:

1. Tags should be appended, we do not want to drop elements, this may block on https://github.com/tendermint/abci/issues/273
2. ValidatorUpdates need to be combined, not dropping redundant elements but rather adding/removing power, the updates should be cumulative
3. Not sure about `GasWanted`, I don't understand why this field is in `Result` anyways. For now this can probably be set to the sum of `GasWanted` for all the individual results.

> Since we abort running the tx on the first message that fails, the result on a failed transaction should just be the result of the first failed message. In the case where the tx successfully executes, we accumulate data/logs/info/tags in the way described above and return the result with code 0.

I think this is OK (to follow ABCI, we have to do this anyways); it would be nice to log that the previous messages succeeded somehow for ease of debugging.

Reply:
Quick question:

In ValidatorUpdates, do we expect address and pubkey to be set correctly. Or do we expect just one to be set correctly like we currently do in x/auth/BaseAccount?

Reply:
>  In ValidatorUpdates, do we expect address and pubkey to be set correctly. Or do we expect just one to be set correctly like we currently do in x/auth/BaseAccount?

Send the address, I think we're getting rid of pubkeys in expectation of large future pubkeys (e.g. BLS).

> If 5 msgs in the same transaction have the same signer, when that transaction is processed, the sequence number for the signer address is only incremented once.
This is because signer addresses are accumulated. (See stdTx.GetSigners() in x/auth/std_tx.go)

I stand corrected, this is an exception (should be the only one I think).